src/model/attributes/relations/MorphTo.ts

```diff
@@ -0,0 +1,140 @@
+import { Query } from '../../../query/Query'
+import { Model } from '../../Model'
+import { Relation } from './Relation'
+
+export class MorphTo extends Relation {
+  /**
+   * The related models.
+   */
+  protected relatedModels: Model[]
+
+  /**
+   * The related model dictionary.
+   */
+  protected relatedTypes: Record<string, Model>
+
+  /**
+   * The field name that contains id of the parent model.
+   */
+  protected morphId: string
+
+  /**
+   * The field name that contains type of the parent model.
+   */
+  protected morphType: string
+
+  /**
+   * The associated key of the child model.
+   */
+  protected ownerKey: string
+
+  /**
+   * Create a new morph-to relation instance.
+   */
+  constructor(
+    parent: Model,
+    relatedModels: Model[],
+    morphId: string,
+    morphType: string,
+    ownerKey: string
+  ) {
+    super(parent, parent)
+
+    this.relatedModels = relatedModels
+    this.relatedTypes = this.createRelatedTypes(relatedModels)
+
+    this.morphId = morphId
+    this.morphType = morphType
+    this.ownerKey = ownerKey
+  }
+
+  protected createRelatedTypes(models: Model[]): Record<string, Model> {
+    const types = {} as Record<string, Model>
+
+    models.forEach((model) => {
+      types[model.$entity()] = model
+    })
+
+    return types
+  }
+
+  /**
+   * Get the type field name.
+   */
+  getType(): string {
+    return this.morphType
+  }
+
+  /**
+   * Get all related models for the relationship.
+   */
+  getRelateds(): Model[] {
+    return this.relatedModels
+  }
+
+  /**
+   * Define the normalizr schema for the relation.
+   */
+  define(schema: Schema): NormalizrSchema {
+    return schema.union(this.relatedModels, (value, parent, _key) => {
+      // Assign missing parent id since the child model is not related back
+      // and `attach` will not be called.
+      const type = parent[this.morphType]
+      const model = this.relatedTypes[type]
+      const key = this.ownerKey || (model.$getKeyName() as string)
+
+      parent[this.morphId] = value[key]
+
+      return type
+    })
+  }
+
+  /**
+   * Attach the relational key to the given record. Since morph-to relationship
+   * doesn't have any foreign key, it would do nothing.
+   */
+  attach(_record: Element, _child: Element): void {
+    return
+  }
+
+  /**
+   * Add eager constraints. Since we do not know the related model ahead of time, we cannot add any
+   * eager constraints.
+   */
+  addEagerConstraints(_query: Query, _models: Collection): void {
+    return
+  }
+
+  /**
+   * Find and attach related children to their respective parents.
+   */
+  match(relation: string, models: Collection, query: Query): void {
+    models.forEach((model) => {
+      let related
+      const type = model[this.morphType]
+      const id = model[this.morphId]
+
+      try {
+        related = query.newQueryWithConstraints(type).find(id)
+      } catch (e) {
+        related = null
+      }
```

Comment:
@kiaking I refactored `match` to catch errors when retrieving related data as we don't know whether or not type and id will be nullable.

For example

```ts
class Image extends Model {
    static entity = 'images'

    @Num(0) id!: number
    @Str('') url!: string
    @Num(0) imageableId!: number
    @Str('') imageableType!: string
    @MorphTo(() => [User, Post], 'imageableId', 'imageableType')
    imageable!: User | Post | null
  }

  class User extends Model {
    static entity = 'users'

    @Num(0) id!: number
    @Str('') name!: string
  }

  class Post extends Model {
    static entity = 'posts'

    @Num(0) id!: number
    @Str('') title!: string
  }
```

Type would default to an empty string an error our when attempting to retrieve the associated model.

Reply:
Matching relations will be based on valid **`id`** and **`type`** values. Remember, each relationship will be responsible for determining what constitutes a valid constraint on its relatables.

In the case of a `morphTo` relation, we know **`id`** can't be `null` or `undefined` and **`type`** must be within the `relatedTypes` boundary. Therefore, the assumption should be made on these values alone to qualify for a related query.

Knowing the invalidity of said values, _trying_ to execute the related query will result in unnecessary thread blocking (since the query will be evaluated until it throws).

I think the following will execute exactly the way we expect it to:
```ts
match(relation: string, models: Collection, query: Query): void {
  const relatedTypes = Object.keys(this.relatedTypes)

  models.forEach((model) => {
    const id = model[this.morphId]

    const type: string | null = relatedTypes.includes(model[this.morphType])
      ? model[this.morphType]
      : null

    isString(type) && !isNullish(id)
      ? model.$setRelation(
          relation,
          query.newQueryWithConstraints(type).find(id)
        )
      : model.$setRelation(relation, null)
  })
}
```

But, of course, `isString` needs to be implemented and some aspects could be broken out into local methods.

Reply:
/cc @kiaking

Reply:
Matching relations will be based on valid **`id`** and **`type`** values. Remember, each relationship will be responsible for determining what constitutes a valid constraint on its relatables.

In the case of a `morphTo` relation, we know **`id`** can't be `null` and **`type`** must be within the `relatedTypes` boundary. Therefore, the assumption should be made on these values alone to qualify for a related query.

Knowing the invalidity of said values, _trying_ to execute the related query will result in unnecessary thread blocking (since the query will be evaluated until it throws).

I think the following will execute exactly the way we expect it to:
```ts
match(relation: string, models: Collection, query: Query): void {
  models.forEach((model) => {
    const id = model[this.morphId]
    const type = model[this.morphType]

    id !== null && this.relatedTypes[type] !== undefined
      ? model.$setRelation(
          relation,
          query.newQueryWithConstraints(type).find(id)
        )
      : model.$setRelation(relation, null)
  })
}
```

We can then remove the optional chaining from [Schema.ts#L120](https://github.com/ryandialpad/vuex-orm-next/blob/c6cef71bff7cd4a17d0ff69e7c065bfb4cb45459/src/schema/Schema.ts#L120) since it is no longer applicable.

Of course, some logic can be broken out into local methods for clarity and readability and the above will still pass all your tests with 100% coverage.

/cc @kiaking

Reply:
For the `match` method, we need to execute `query.newQueryWithConstraints` outside of the `forEach` loop. Because if we have 100 models here, it will create 100 query instances, which is unneeded.

I think it would be better to create dictionary before doing `forEach` loop like other relationships. In `MorphTo` relation, we could create dictionary for all possible related models first.

I haven't tested the code so it might fail but it should look something like this 👀 

```ts
interface DictionaryByEntities {
  [entity: string]: {
    [id: string]: Model
  }
}

match(relation: string, models: Collection, query: Query): void {
  // Create dictionary that contains relationships.
  const dictionary = this.buildDictionary(query)

  models.forEach((model) => {
    const type = model[this.morphType]
    const id = model[this.morphId]

    const related = dictionary[type][id] ?? null

    model.$setRelation(relation, related)
  })
}

buildDictionary(query: Query, models: Collection): DictionaryByEntities {
  const keys = this.getKeysByEntity(models)

  const dictionary = {} as DictionaryByEntities

  for (const entity in keys) {
    const model = this.relatedTypes[entity]

    // If we can't find a model here, it means the user did not provide model
    // that corresponds with the type. We can throw here to inform users.
    assert(!!model, [
      `Trying to load "morph to" relation of \`${entity}\` but the model ` +
      `could not be found.`
    ])

    const ownerKey = this.ownerKey || model.$getKeyName()

    const results = query
      .newQueryWithConstraints(entity)
      .whereIn(owenerKey, keys[entity])
      .get()

    dictionary[entity] = results.reduce<Record<string, Model>>((dic, result) => {
      dic[result[ownerKey]] = result

      return dic
    }, {})
  }

  return dictionary
}

getKeysByEntity(models: Collection): Record<string, (string | number)[]> {
  return models.reduce<Record<string, (string | number)[]>>((keys, model) => {
    const type = model[this.morphType]
    const id = model[this.morphId]

    // TODO: Check if both type and id is valid. If not, we should just ignore it.

    if (!keys[type]) {
      keys[type] = []
    }

    keys[type].push(id)

    return keys
  }, {})
}
```

The idea is to first loop through all the models, and gather ids for the related models. Then fetch all related models and build dictionary. Finally, just access the dictionary to see if we can find corresponding relationship.

I think we should throw if user provide non existing type. For example, if user tries to load relationship to `{ imageableType: 'posts' }`, and if user did not provide Post model to the `MorphTo` method, `@MorphTo(() => [User])`, then we throw saying that model named `posts` is not registered.

Reply:
Thanks for the feedback and discussion @cuebit and @kiaking!!!